proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+This would primarily be of value in cross-compilation scenarios.
+
+In the interests of keeping the work manageable, changes to Cabal are out of
+scope for the current proposal, but we believe this proposal lays a foundation
+for future work to improve Cabal's cross-compilation support.
+
+
+Imports with explicit level numbers
+###################################
+
+The current proposal permits imports only at levels -1, 0 or 1. This means it is
+not possible to introduce a binding for use in a splice contained within another
+splice, which would require it to be at level -2.  (Note that nested quotes are
+in any case not supported in GHC due to a separate restriction.)
+
+An alternative would be to allow even finer grained control of splice imports so
+that usage at level -2 or lower could be distinguished. This could be useful in
+some cross-compilation situations. This is the approach suggested in the `Stage
+Hygiene for Template Haskell proposal
+<https://github.com/ghc-proposals/ghc-proposals/pull/243>`_.
+
+The syntax in this proposal could be extended in a natural way to allow for this by adding an optional
+integer component which specifies precisely what level the imported names should be allowed at::
+
+    -- Can be used at level -1
+    import splice 1 A
+    -- Can be used at level -2
+    import splice 2 A
+
+Practically, by far the most common situation is a single level of splices, so in the interests
+of reducing complexity we do not propose supporting this at present.
+
+
+Inferring splice/quote imports based on usage
+#############################################
+
+Since our proposed approach has the type-checker verify that usage of ``splice``
+or ``quote`` imports is correct, it may be possible in principle to infer where
+``splice`` or ``quote`` keywords are needed, based on usage inside a module.
+However, this would compromise the principle that the build system can discover
+the dependencies for a module just by looking at the import list in the module
+header. Achieving the performance benefits of our proposed approach would
+involve significant technical complexity (as the compiler would need to
+partially type-check a module, then suspend compilation of that module while it
+compiles those of its dependencies determined to be required for further
+type-checking).
+
+Given that the ``splice`` and ``quote`` annotations are useful for human readers
+understanding how code is staged, it seems worthwhile to make them explicit.
+
+Of course, nothing prevents development of a tool that helps users insert
+``splice`` and ``quote`` annotations into their modules as part of a migration
+to using ``ExplicitLevelImports``.
+
+
+Syntactic alternatives
+######################
+
+* Using a pragma rather than a syntactic modifier would fit in better with
+  how ``SOURCE`` imports work and make writing backwards compatible code easier::
+
+    import {-# SPLICE #-} B
```

Comment:
I think I would like that, because it meant that adding splice import annotations are completely backwards compatible, **if** it weren't for the fact that [unknown extension flags cause GHC to error](https://gitlab.haskell.org/ghc/ghc/-/issues/23858). Grrrr